sdks/python/apache_beam/runners/interactive/background_caching_job.py

```diff
@@ -0,0 +1,101 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Module to build and run background caching job.
+
+For internal use only; no backwards-compatibility guarantees.
+
+A background caching job is a job that caches events for all unbounded sources
+of a given pipeline. With Interactive Beam, one such job is started when a
+pipeline run happens (which produces a main job in contrast to the background
+caching job) and meets the following conditions:
+
+  #. The pipeline contains unbounded sources.
+  #. No such background job is running.
+  #. No such background job has completed successfully and the cached events are
+     still valid (invalidated when unbounded sources change in the pipeline).
+
+Once started, the background caching job runs asynchronously until it hits some
+cache size limit. Meanwhile, the main job and future main jobs from the pipeline
+will run using the deterministic replay-able cached events until they are
+invalidated.
+"""
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+from apache_beam import runners
+from apache_beam.runners.interactive import interactive_environment as ie
+
+
+def attempt_to_run_background_caching_job(runner, user_pipeline, options=None):
+  """Attempts to run a background caching job for a user-defined pipeline.
+
+  If a background caching job is started, return the pipeline result. Otherwise,
+  return None.
+  The pipeline result is automatically tracked by Interactive Beam in case
+  future cancellation/cleanup is needed.
+  """
+  if is_background_caching_job_needed(user_pipeline):
+    # Cancel non-terminal jobs if there is any before starting a new one.
+    attempt_to_cancel_background_caching_job(user_pipeline)
+    # TODO(BEAM-8335): refactor background caching job logic from
+    # pipeline_instrument module to this module and aggregate tests.
+    from apache_beam.runners.interactive import pipeline_instrument as instr
+    runner_pipeline = beam.pipeline.Pipeline.from_runner_api(
+        user_pipeline.to_runner_api(use_fake_coders=True),
+        runner,
+        options)
+    background_caching_job_result = beam.pipeline.Pipeline.from_runner_api(
+        instr.pin(runner_pipeline).background_caching_pipeline_proto(),
+        runner,
+        options).run()
+    ie.current_env().set_pipeline_result(user_pipeline,
+                                         background_caching_job_result,
+                                         is_main_job=False)
+
+def is_background_caching_job_needed(user_pipeline):
+  """Determines if a background caching job needs to be started."""
+  background_caching_job_result = ie.current_env().pipeline_result(
+      user_pipeline, is_main_job=False)
+  from apache_beam.runners.interactive import pipeline_instrument as instr
+  # Checks if the pipeline contains unbounded sources.
+  return (instr.has_unbounded_sources(user_pipeline) and
```

Comment:
This may not be in the scope of this PR, but we also need to let the user specify their own sources to be cached, because the user may have a large bounded source that they want to capture a small segment of.

Reply:
Got it, I'll make it a function in `background_caching_job` called `has_source_to_cache` and move the logic there. Future work can be done there too.